sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -274,7 +274,36 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
 
   val diskBlockObjectWriters = new mutable.HashMap[Int, (Int, DiskBlockObjectWriter)]()
 
+  /**
+   * Simple wrapper that tracks the time spent iterating the given iterator.
+   */
+  class TimeTrackingIterator(delegate: Iterator[Product2[K, V]]) extends Iterator[Product2[K, V]] {
+    private var iterateTimeNs: Long = 0L
+
+    override def hasNext: Boolean = {
+      val start = System.nanoTime()
+      val ret = delegate.hasNext
+      iterateTimeNs += System.nanoTime() - start
+      ret
+    }
+
+    override def next(): Product2[K, V] = {
+      val start = System.nanoTime()
+      val ret = delegate.next
+      iterateTimeNs += System.nanoTime() - start
+      ret
+    }
+
+    def getIterateTimeNs: Long = iterateTimeNs
+  }
+
   override def write(records: Iterator[Product2[K, V]]): Unit = {
+    // Iterating the `records` may involve some heavy computations.
+    // TimeTrackingIterator is used to track how much time we spend for such computations.
+    write(new TimeTrackingIterator(records))
+  }
+
+  def write(records: TimeTrackingIterator): Unit = {
```

Comment:
nit, lets mark this `private`, I like the addition of the new method.

Reply:
Made this and a couple of others private.